tests/FSharp.Compiler.ComponentTests/CompilerService/AsyncMemoize.fs

```diff
@@ -101,6 +101,8 @@ let ``We can cancel a job`` () =
 
         Assert.Equal<(JobEvent * int) array>([| Started, key |], eventLog |> Seq.toArray )
 
+        do! Task.Delay 100
```

Comment:
Is there a way to phrase this with explicit gates within the computations so that we know when execution has progressed into the tasks, and then the tasks block on the cancellation of the CTS? Unless there's some pretty mad shenanigans somewhere, these tests are inherently flaky (#TimeIsNotASynchronisationPrimitive). I would expect at some point the failwith on line 80 to be hit, if nothing else, when during some test run the task scheduler happens not to schedule the main thread (which performs the cancellation) for one second; such coincidences are inevitable when running large test suites many times.

Reply:
Yeah, you're right. I should figure out how to do it without arbitrary time delays. Probably use the events somehow to progress the test.

Reply:
(There's nothing wrong with a helper function that spins waiting for a sequence to contain n elements, for example, although since one must tediously ensure the collection is safe to read and write concurrently, for tests I usually wrap it up in a type that controls the accesses made to it and which has a `Freeze : unit -> 'a list` method.)